### PR TITLE
Added storing attribute values as post_excerpt for variations.

### PR DESCRIPTION
**Pull request by peterfabian**
This allows searching for variations by their attributes without cluttering the Variation's title.
If the target version will be different than 3.5.3, the db routine needs to be updated.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/wc-admin/issues/831 .

### How to test the changes in this Pull Request:

1. Create a variable product with 4 attributes.
2. Create variations for the product.
3. Try to search for the variations of the parent product by attribute value, e.g. color via REST API: `https://local.wordpress.test/wp-json/wc/v3/products/1168/variations/?search=black`
4. Old version won't work, new version should work fine.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### Changelog entry

> Added storing attribute values as post_excerpt for variations to support easier searching for variations.

### CODE REVIEW

**Review by claudiosanches — CHANGES_REQUESTED** [review]
`post_excerpt` is a default field on WordPress, this probably may generate backwards compatibility issues, since plugins or themes may change it, and for this PR, seems like will be completely replaced with this new content.

Also it's not a patch or a fix, so sounds big to be included into 3.5.3.
This is something that will require more conversion about.

**Comment by claudiosanches** [comment]
> Try to search for the variations of the parent product by attribute value, e.g. color via REST API: https://local.wordpress.test/wp-json/wc/v3/products/1168/variations/?search=black

In this case we should allow filter by attributes, or even improve it.

**Comment by peterfabian** [comment]
Agreed that it needs more discussion, I created this PR to kind of kick off the topic and discuss it.

I agree that if included, it should probably be in 3.6, not in 3.5.3, as it is a breaking change. 
From performance point of view, I wanted to avoid implementing it using meta queries (which would currently require 1 JOIN per attribute). 

Maybe a better way to implement this would be to add the same info to one meta field instead of `post_excerpt` and then it would require only one join with the meta table? What do you think?

**Comment by claudiosanches** [comment]
@peterfabian we probably can go with `post_excerpt`, but we need to communicate developers before (dev blog and so on), I agree with you about performance, `post_excerpt` is something easy to include in the search. For sure a meta field should be easy and safer, but not so good in performance like `post_excerpt`.

**Comment by mikejolley** [comment]
The attributes being missing from the post title was to avoid titles being super long on the cart page - in those cases it switches to a list view.

Advantage of using post excerpt may be that we can fill those attributes and store them at all times even when long, and then can decide to switch to a list view on cart page based on excerpt length. So that sounds good.

Do we need a DB routine though? Thats going to be super resource intensive. Could we do it in the data store on access instead?

**Comment by peterfabian** [comment]
@mikejolley Do you think we should check on each variable product access in data store if it's not empty and if it is, then update? I'm just a bit afraid it's a bit wasteful to have the code there for each access. 
How about an Action Scheduler queue for these updates?

**Comment by mikejolley** [comment]
@peterfabian That was my suggestion. Realistically it would then have to generate for each variable product on each shop page/admin once only, only when needed. Like lazy loading..

If we need to do a DB routine we could consolodate this in with the new sorting table routine..

**Comment by peterfabian** [comment]
Updated according to the suggestion. I took the conservative approach to only overwrite if the field is empty, so that other info stored there would not get overwritten. On the other hand, it might mean the search would not work correctly. Please let me know if you think we should overwrite in case of different value, similarly to how name is handled above.

Thinking about it now, maybe excerpt is not the best place for this string. One reason is that simple products can also have attributes, which should be also searchable, but excerpt is already used for `short_description`. Should we find a different place for this info?

**Review by mikejolley — COMMENTED** [review]
I've suggested adding a new prop for this with getter/setter. The reason for this is we can potentially use this data in the cart and in the `get_formatted_name` to save recreating this string of values every time we need to display the string of attributes. Suggested `summary`.

If this is beyond what you had in mind, maybe we can instead consider a lookup table but let's chat about that first.

**Comment by peterfabian** [comment]
Thanks for the review, updated. 
Additional test cases to check:
- create a new attribute for variable product -> refresh product page -> check database that excerpt got updated
- change attribute value or name -> refresh product -> check database that excerpt got updated

File: includes/class-wc-product-variation.php [hunk 1]
```diff
@@ -413,6 +416,34 @@ public function get_catalog_visibility( $context = 'view' ) {
...
+	 * By default, attribute summary contains comma-delimited 'attribute_name: attribute_value' pairs for all attributes.
+	 *
+	 * @param string $context What the value is for. Valid values are view and edit.
+	 *
+	 * @since 3.6.0
+	 * @return string
+	 */
+	public function get_attribute_summary( $context = 'view' ) {
+		$attribute_summary = $this->get_prop( 'attribute_summary', $context );
+		return apply_filters( $this->get_hook_prefix() . 'attribute_summary', $attribute_summary, $this );
```
Comment (mikejolley) [inline includes/class-wc-product-variation.php]: Why are you manually calling the filter here? `get_prop` handles filtering.

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php [hunk 2]
```diff
@@ -85,6 +87,19 @@ public function read( &$product ) {
...
 		}
 
+		/**
+		 * If the attribute summary is not in sync, update here. Used when searching for variations by attribute values.
+		 * This is meant to also cover the case when global attribute name or value is updated, then the attribute summary is updated
+		 * for respective products when they're read.
+		 */
+		$new_attribute_summary = $this->generate_attribute_summary( $product );
+
+		if ( $new_attribute_summary !== $post_object->post_excerpt ) {
```
Comment (mikejolley) [inline includes/data-stores/class-wc-product-variation-data-store-cpt.php]: Can we combine the post_title update and this update into one update, so it's not called twice in some circumstances?

**Review by mikejolley — COMMENTED** [review]
2 comments with feedback needed. Once resolved I'll do some tests.

Does this also need unit tests?

File: includes/data-stores/class-wc-product-variation-data-store-cpt.php (hunk 2 above)
Comment (peterfabian) [reply includes/data-stores/class-wc-product-variation-data-store-cpt.php]: Good idea, updated.

File: includes/class-wc-product-variation.php (hunk 1 above)
Comment (peterfabian) [reply includes/class-wc-product-variation.php]: You're right, thx. Got confused by other getters running their filters.

**Comment by peterfabian** [comment]
@mikejolley Updated and added tests, hope it looks better now.

**Review by mikejolley — APPROVED** [review]
Tested and working 👍 Looks good now.

**Pull request merged**